Sources/StreamChat/Workers/Background/DatabaseCleanupUpdater.swift

```diff
@@ -31,48 +31,91 @@ class DatabaseCleanupUpdater: Worker {
+                                .forEach { $0.resetLocalData() }
+                            
+                            // Unlink all channels from a query
+                            queryDTO.channels.removeAll()
+                            
+                            // Link channels from first page to query
+                            for channel in firstPage.channels {
+                                try session.saveChannel(
+                                    payload: channel,
+                                    query: queryDTO.asModel()
+                                )
+                            }
+                        case let .failure(error):
+                            log.error("Failed to re-fetch channel list query \(queryHash) \(error)")
+                        }
+                    }
+                }
             }
         }
     }
-    
-    /// Finds the existing channel list queries in the database and refetches them.
-    func refetchExistingChannelListQueries() {
+}
+
+// MARK: - Private
+
+private extension DatabaseCleanupUpdater {
+    func fetchLocalQueries(completion: @escaping ([ChannelListQuery]) -> Void) {
         let context = database.backgroundReadOnlyContext
-        context.perform { [weak self] in
-            do {
-                let queriesDTOs = try context.fetch(
-                    NSFetchRequest<ChannelListQueryDTO>(
-                        entityName: ChannelListQueryDTO.entityName
-                    )
-                )
-                let queries: [ChannelListQuery] = try queriesDTOs.map {
-                    try $0.asChannelListQuery()
-                }
-                queries.forEach {
-                    self?.channelListUpdater.update(channelListQuery: $0) { result in
-                        if case let .failure(error) = result {
-                            log.error("Internal error. Failed to update ChannelListQueries for the new channel: \(error)")
-                        }
-                    }
-                }
-            } catch {
-                log.error("Internal error: Failed to fetch [ChannelListQueryDTO]: \(error)")
+        context.perform {
+            let queries = context
+                .loadChannelListQueries()
+                .compactMap { $0.asModel() }
+            
+            completion(queries)
+        }
+    }
+    
+    func fetchFirstPage(
+        of queries: [ChannelListQuery],
+        completion: @escaping ([String: Result<ChannelListPayload, Error>]) -> Void
+    ) {
+        let group = DispatchGroup()
+        let semaphore = DispatchSemaphore(value: 1)
+        
+        var results: [String: Result<ChannelListPayload, Error>] = [:]
+        for query in queries {
+            group.enter()
+            
+            log.debug("Will fetch first page of channels for \(query)")
+            channelListUpdater.fetch(query) {
+                semaphore.wait()
```

Comment:
Instead of using a semaphore, we can create a temp serial queue and use it to access:
```
let queue = DispatchQueue("dict-access")
....
queue.async { results[query.queryHash] = queryHash }
....
```

Reply:
for some async operations we could carry in QueuedExecutor 
from my draft PR on Members Pagination

Reply:
for syncronization of dictionary, I would also recommend @b-onc  approach, 
cause serial queues is general mechanism recommended by Apple to handle concurrency syncronizations